accepted/2021/Globalization/GlobalizationSize.md

```diff
@@ -0,0 +1,81 @@
+
+### ICU4X
+
+It is interesting enough, there is a newly launched open-source project called [ICU4X](https://github.com/unicode-org/icu4x#icu4x----) for introducing globalization support libraries (similar to ICU). ICU4X has the exact goals we need to achieve with our concerned scenarios.
+
+```
+ICU4X will provide an ECMA-402-compatible API surface in the target client-side platforms, including the web platform, iOS, Android, WearOS, WatchOS, Flutter, and Fuchsia, supported in programming languages including Rust, JavaScript, Objective-C, Java, Dart, and C++.
+
+The design goals of ICU4X are:
+- Small and modular code
+- Pluggable locale data
+- Availability and ease of use in multiple programming languages
+- Written by i18n experts to encourage best practices
+```
+
+That is exactly what we need (or what we are trying to achieve).
+Eric Erhardt already did some investigation and arranged a meeting with the project committee to learn more about the project and the status of the project. It was very helpful getting in touch and learning more about this project. It is a promising project managed by experts and using the CLDR data which means it still sticking with the standards.
+
+The catch here is ICU4X still in the early stages and still not providing all functionality needed by .NET for globalization support. The project committee is willing to get and prioritize requests from different parties. We already communicated in the meeting what functionality .NET currently using from ICU4C and want to see that supported in ICU4X. For example, collation support is one of the topics we brought to the committee's attention.
+
+### Plan Suggestion
+
+ICU4X is the promising path we should pursue to address all concerned scenarios. We can wait a little bit more to get the needed missing functionality implemented in the project and integrate it to .NET as we did with ICU4C.
+
```

Comment:
This reason for doing this is performance / size. As such, this should be heavily backed by numbers. I think the priority should be to get estimates for where we may land by doing this work.

- The code size for full .NET globalization support: current stripped ICU vs. estimate for ICU4X
- The data size for full .NET globalization support: current stripped ICU vs. estimate for ICU4X

- The code modularity: pick a flagship scenario for code modularity, and show current vs. estimate for ICU4X
- The data modularity: pick a flagship scenario for data modularity, and show current vs. estimate for ICU4X

Reply:
> estimate for ICU4X

Unfortunately, this is close to impossible at the current time. The features we need just aren't there. So getting any sort of size estimate is purely guessing. We would need to implement the features first.

Reply:
Do they have some interesting large chunk implemented that we can triangulate with the same chunk of classic ICU?

It is hard to tell whether ICU4X is promissing path without any numbers.

Reply:
Looking through the features on their [releases](https://github.com/unicode-org/icu4x/releases), DateTime formatting might be a candidate (that's what I was trying to prototype against a few months ago), but we don't really use ICU for the formatting logic, just to get the data.

Reply:
Yeah, I think it might be possible to get numbers for *something* but I'm not sure they would be very meaningful at this point. The project, given its current state, is definitely inherently speculative.

Reply:
I'll add a section at the end of the documents about what different things we need to know to make form the decision.

Reply:
Also, I would say that the benefit has to be large (2x or more) to make this endevour worth it.

For example, do we believe that ICU4X will make the total (code+data) size footprint for full .NET globalization support to be at least 2x better compared to what we can get with stripped down ICU today?

Reply:
> For example, do we believe that ICU4X will make the total (code+data) size footprint for full .NET globalization support to be at least 2x better compared to what we can get with stripped down ICU today?

I doubt this will be the case. I think this is not the goal we are shooting for. We are trying to target the scenarios that need smaller disk footprint (mobile and wasm platforms). Teh idea is to use trimmed data and code to address the issues on these platforms. Just wondering why, we need to have this as a gool? I can understand this be a goal for mobile and wasm platforms but not for full globalization support case.

Reply:
My hypothesis is that there are two main scenarios: full globalization (app with UI) and no globalization (microservice or command line app). It is why I think we should focus on full globalization and no globalization (invariant mode).

Once you start to slice and dice the globalization, it gets really difficult to explain to what app authors should do for their app with UI as we have discussed in other fork. I expect that most people will just go with full globalization to avoid dealing with the complexity.

So the only thing left are niche scenarios that want to micro-manage the globalization footprint. We satisfly some of this with ICU by slicing the data today. I do not think we need to boil the ocean for these.

Reply:
> My hypothesis is that there are two main scenarios: full globalization (app with UI) and no globalization (microservice or command line app). It is why I think we should focus on full globalization and no globalization (invariant mode).

@jkotas is Blazor in the first category for you?

@danroth27 can you share more about the globalization scenarios you expect from Blazor users? How many will care enough about size to want to have less than full globalization? And for those, how many would be satisfied with minimal/no globalization support vs configuring some custom subset if available.

Reply:
> @jkotas is Blazor in the first category for you?

Yes, I would expect typical Blazor app to be in the first category.

Reply:
Hi, this is Shane from the ICU4X team. I'm happy to get you ballpark figures for code+data size of ICU4C vs ICU4X.

Reply:
Mozilla I18n Team is currently running an early signal performance testing of ICU4C vs ICU4X for our JS engine and internal needs. We hope to be able to provide some integration experience and performance/memory numbers within a month. If you're interested, we will be hosting a Deep Dive session in July where we plan to present the results.

Reply:
Thanks! And, to follow up from above, I posted some ballpark code size figures here:

https://github.com/unicode-org/icu4x/discussions/788

Reply:
> @danroth27 can you share more about the globalization scenarios you expect from Blazor users? How many will care enough about size to want to have less than full globalization? And for those, how many would be satisfied with minimal/no globalization support vs configuring some custom subset if available.

Web developers are generally very sensitive to app download size and front end frameworks compete to be as small as possible. So in general I believe web devs will want just enough functionality to run their app and nothing more. There is a subset of web scenarios where app download size is less important, but it is the minority scenario. I suspect there are a significant number of web scenarios that could get by with just invariant mode and some light weight globalization helpers based on browser platform APIs, but @jkotas is probably right that the majority of web apps need broad globalization support, at least enough to handle their target audiences.